datafusion/core/tests/user_defined/user_defined_window_functions.rs

```diff
@@ -0,0 +1,548 @@
+        "| 2 | j | 6   | 2                  |",
+        "+---+---+-----+--------------------+",
+    ];
+    assert_batches_eq!(
+        expected,
+        &execute(&ctx, UNBOUNDED_WINDOW_QUERY).await.unwrap()
+    );
+    // evaluated on three distinct batches
+    assert_eq!(test_state.evaluate_all_called(), 2);
+}
+
+/// Basic user defined window function with bounded window
+#[tokio::test]
+async fn test_udwf_bounded_window_ignores_frame() {
+    let test_state = TestState::new();
+    let TestContext { ctx, test_state } = TestContext::new(test_state);
+
+    // Since the UDWF doesn't say it needs the window frame, the frame is ignored
+    let expected = vec![
+        "+---+---+-----+--------------------+",
+        "| x | y | val | odd_counter(t.val) |",
+        "+---+---+-----+--------------------+",
+        "| 1 | a | 0   | 1                  |",
+        "| 1 | b | 1   | 1                  |",
+        "| 1 | c | 2   | 1                  |",
+        "| 2 | d | 3   | 2                  |",
+        "| 2 | e | 4   | 2                  |",
+        "| 2 | f | 5   | 2                  |",
+        "| 2 | g | 6   | 2                  |",
+        "| 2 | h | 6   | 2                  |",
+        "| 2 | i | 6   | 2                  |",
+        "| 2 | j | 6   | 2                  |",
+        "+---+---+-----+--------------------+",
+    ];
+    assert_batches_eq!(
+        expected,
+        &execute(&ctx, BOUNDED_WINDOW_QUERY).await.unwrap()
+    );
+    // evaluated on 2 distinct batches (when x=1 and x=2)
+    assert_eq!(test_state.evaluate_called(), 0);
+    assert_eq!(test_state.evaluate_all_called(), 2);
+}
+
+/// Basic user defined window function with bounded window
+#[tokio::test]
+async fn test_udwf_bounded_window() {
+    let test_state = TestState::new().with_uses_window_frame();
+    let TestContext { ctx, test_state } = TestContext::new(test_state);
+
+    let expected = vec![
+        "+---+---+-----+--------------------+",
+        "| x | y | val | odd_counter(t.val) |",
+        "+---+---+-----+--------------------+",
+        "| 1 | a | 0   | 1                  |",
+        "| 1 | b | 1   | 1                  |",
+        "| 1 | c | 2   | 1                  |",
+        "| 2 | d | 3   | 1                  |",
+        "| 2 | e | 4   | 2                  |",
+        "| 2 | f | 5   | 1                  |",
+        "| 2 | g | 6   | 1                  |",
+        "| 2 | h | 6   | 0                  |",
+        "| 2 | i | 6   | 0                  |",
+        "| 2 | j | 6   | 0                  |",
+        "+---+---+-----+--------------------+",
+    ];
+    assert_batches_eq!(
+        expected,
+        &execute(&ctx, BOUNDED_WINDOW_QUERY).await.unwrap()
+    );
+    // Evaluate is called for each input rows
+    assert_eq!(test_state.evaluate_called(), 10);
+    assert_eq!(test_state.evaluate_all_called(), 0);
+}
+
+/// Basic stateful user defined window function
+#[tokio::test]
+async fn test_stateful_udwf() {
```

Comment:
I couldn't think of anything else to test with a stateful function other than `update_state` is called. I would be interested in anyone else's opinions in this matter